### PR TITLE
Issue/128

### LINKED ISSUE #128: Create Solutions (MultipleChoice and Wordsalad)
# Description

Create Endpoint and Controller for MultipleChoice and Wordsaladsolutions

- create 
- update
- delete
- get
- check
... solutions

---

remove updateMultipleChoice (unnecessary)

Delete: two Pathvariables : taskId and solution

### PR DESCRIPTION
**Pull request by PAException**
## Description

Closes #128 

## Type of change

- [x] Bug fix
- [x] Feature request
- [ ] Documenting code

## Checklists

- [x] I referenced the according Issue
  - [x] I created an Issue
  - [x] I put a link in the Section [Related Issue](#related-issue)
- [ ] I tested my code with multiple browsers
- [x] My commit messages are good and understandable

### CODE REVIEW

File: .docs/solutionTemplates/controller/solution/SolutionNAMEController.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import de.themorpheus.edu.taskservice.database.repository.solution.SolutionNAMERepository;
+import de.themorpheus.edu.taskservice.endpoint.dto.solution.NAME.CheckSolutionNAMEDTO;
+import de.themorpheus.edu.taskservice.endpoint.dto.solution.NAME.CreateSolutionNAMEDTO;
+import de.themorpheus.edu.taskservice.endpoint.dto.solution.NAME.UpdateSolutionNAMEDTO;
+import de.themorpheus.edu.taskservice.util.ControllerResult;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class SolutionNAMEController {
```
Comment (steve-hb) [inline .docs/solutionTemplates/controller/solution/SolutionNAMEController.java]: CamelCase

File: .docs/solutionTemplates/controller/solution/SolutionNAMEController.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+	@Autowired private SolutionNAMERepository solutionNAMERepository;
```
Comment (steve-hb) [inline .docs/solutionTemplates/controller/solution/SolutionNAMEController.java]: CamelCase

File: src/main/java/de/themorpheus/edu/taskservice/controller/solution/WordsaladSolutionController.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class WordsaladSolutionController {
+
+	@Autowired private SolutionWordsaladRepository solutionWordsaladRepository;
+
+	public ControllerResult<WordsaladSolutionModel> create(CreateWordsaladSolutionDTO createWordsaladSolutionDTO) {
+		if (this.solutionWordsaladRepository.findById(createWordsaladSolutionDTO.getTaskId()).isPresent()) return ControllerResult.of(Error.ALREADY_EXISTS);
```
Comment (steve-hb) [inline src/main/java/de/themorpheus/edu/taskservice/controller/solution/WordsaladSolutionController.java]: Create method in controller: exists(int taskId)

File: src/main/java/de/themorpheus/edu/taskservice/controller/solution/WordsaladSolutionController.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+		WordsaladSolutionModel wordsaladSolutionModel = new WordsaladSolutionModel();
```
Comment (steve-hb) [inline src/main/java/de/themorpheus/edu/taskservice/controller/solution/WordsaladSolutionController.java]: Use contructor

File: src/main/java/de/themorpheus/edu/taskservice/controller/solution/WordsaladSolutionController.java (hunk 3 above)
Comment (steve-hb) [reply src/main/java/de/themorpheus/edu/taskservice/controller/solution/WordsaladSolutionController.java]: also use repository method for existence

File: src/main/java/de/themorpheus/edu/taskservice/controller/solution/WordsaladSolutionController.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+		wordsaladSolutionModel.setTaskId(createWordsaladSolutionDTO.getTaskId());
+		return ControllerResult.of(this.solutionWordsaladRepository.save(wordsaladSolutionModel));
+	}
+
+	public ControllerResult<WordsaladSolutionModel> check(CheckWordsaladSolutionDTO checkWordsaladSolutionDTO) {
+		Optional<WordsaladSolutionModel> solutionWordsaladModels = this.solutionWordsaladRepository.findById(checkWordsaladSolutionDTO.getTaskId());
+		if (!solutionWordsaladModels.isPresent()) return ControllerResult.of(Error.NOT_FOUND);
+
+		WordsaladSolutionModel wordsaladSolutionModel = solutionWordsaladModels.get();
+		if (wordsaladSolutionModel.getSolution().equals(checkWordsaladSolutionDTO.getSolution())) {
```
Comment (steve-hb) [inline src/main/java/de/themorpheus/edu/taskservice/controller/solution/WordsaladSolutionController.java]: equalsIgnoreCase

File: src/main/java/de/themorpheus/edu/taskservice/controller/solution/WordsaladSolutionController.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+		} else {
+			return ControllerResult.of(Error.WRONG_ANSWER);
+		}
+	}
+
+	public ControllerResult<WordsaladSolutionModel> update(UpdateWordsaladSolutionDTO updateWordsaladSolutionDTO) {
+		WordsaladSolutionModel wordsaladSolutionModel = new WordsaladSolutionModel();
+		wordsaladSolutionModel.setSolution(updateWordsaladSolutionDTO.getSolution());
+		wordsaladSolutionModel.setTaskId(updateWordsaladSolutionDTO.getTaskId());
+		return ControllerResult.of(wordsaladSolutionModel);
```
Comment (steve-hb) [inline src/main/java/de/themorpheus/edu/taskservice/controller/solution/WordsaladSolutionController.java]: where do u update this in the db?

File: src/main/java/de/themorpheus/edu/taskservice/controller/solution/WordsaladSolutionController.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+	}
+
+	public ControllerResult<WordsaladSolutionModel> delete(int taskId) {
+		if (!this.solutionWordsaladRepository.findById(taskId).isPresent()) return ControllerResult.of(Error.NOT_FOUND);
```
Comment (steve-hb) [inline src/main/java/de/themorpheus/edu/taskservice/controller/solution/WordsaladSolutionController.java]: use exists method

File: src/main/java/de/themorpheus/edu/taskservice/controller/solution/WordsaladSolutionController.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+		this.solutionWordsaladRepository.deleteById(taskId);
+		return ControllerResult.empty();
+	}
+
+	public ControllerResult<GetWordsaladSolutionDTO> get(int taskId) {
+		Optional<WordsaladSolutionModel> solutionWordsaladModels = this.solutionWordsaladRepository.findById(taskId);
+		if (!solutionWordsaladModels.isPresent()) return ControllerResult.of(Error.NOT_FOUND);
```
Comment (steve-hb) [inline src/main/java/de/themorpheus/edu/taskservice/controller/solution/WordsaladSolutionController.java]: exists method

File: src/main/java/de/themorpheus/edu/taskservice/controller/solution/WordsaladSolutionController.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+		WordsaladSolutionModel wordsaladSolutionModel = solutionWordsaladModels.get();
+		List<Character> characters = new ArrayList<>();
```
Comment (steve-hb) [inline src/main/java/de/themorpheus/edu/taskservice/controller/solution/WordsaladSolutionController.java]: new ArrayList<>(wordsaladSolutionModel.getSolution().length());

File: src/main/java/de/themorpheus/edu/taskservice/database/repository/solution/SolutionRepository.java [hunk 10]
```diff
@@ -0,0 +1,12 @@
+package de.themorpheus.edu.taskservice.database.repository.solution;
+
+import de.themorpheus.edu.taskservice.database.model.solution.SolutionModel;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface SolutionRepository extends JpaRepository<SolutionModel, Integer> {
+
```
Comment (steve-hb) [inline src/main/java/de/themorpheus/edu/taskservice/database/repository/solution/SolutionRepository.java:10]: Too many empty lines

File: src/main/java/de/themorpheus/edu/taskservice/database/repository/solution/SolutionWordsaladRepository.java [hunk 11]
```diff
@@ -0,0 +1,12 @@
+package de.themorpheus.edu.taskservice.database.repository.solution;
+
+import de.themorpheus.edu.taskservice.database.model.solution.WordsaladSolutionModel;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface SolutionWordsaladRepository extends JpaRepository<WordsaladSolutionModel, Integer> {
+
```
Comment (steve-hb) [inline src/main/java/de/themorpheus/edu/taskservice/database/repository/solution/SolutionWordsaladRepository.java:9]: Too many empty lines

File: src/main/java/de/themorpheus/edu/taskservice/endpoint/solution/WordsaladSolutionEndpoint.java [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class WordsaladSolutionEndpoint {
+
+	@Autowired
+	WordsaladSolutionController wordsaladSolutionController;
```
Comment (steve-hb) [inline src/main/java/de/themorpheus/edu/taskservice/endpoint/solution/WordsaladSolutionEndpoint.java]: Same line

File: src/main/java/de/themorpheus/edu/taskservice/util/ControllerResult.java [hunk 13]
```diff
@@ -82,7 +82,7 @@ public boolean isEmpty() {
 	 */
 	@SuppressWarnings({"unchecked", "rawtypes"})
 	public static <T> ControllerResult<T> ret(ControllerResult instance) {
-		if (instance.isErrorPresent()) throw new IllegalArgumentException("No error present in controller result");
+		if (instance.isErrorPresent()) throw new IllegalArgumentException("No error present in de.themorpheus.edu.taskservice.controller result");
```
Comment (steve-hb) [inline src/main/java/de/themorpheus/edu/taskservice/util/ControllerResult.java]: reset

**Review by steve-hb — CHANGES_REQUESTED** [review]

File: .docs/solutionTemplates/de/themorpheus/edu/taskservice/controller/solution/GenericSolutionController.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+import de.themorpheus.edu.taskservice.util.ControllerResult;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class GenericSolutionController {
+
+	@Autowired private GenericSolutionRepository genericSolutionRepository;
+
+	public ControllerResult<GenericSolutionModel> create(CreateGenericSolutionDTO createGenericSolutionDTO) {
```
Comment (steve-hb) [inline .docs/solutionTemplates/de/themorpheus/edu/taskservice/controller/solution/GenericSolutionController.java]: create -> createGenericSolutionController
change this everywhere

File: .docs/solutionTemplates/de/themorpheus/edu/taskservice/controller/solution/GenericSolutionController.java [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+		return null;	//TODO: Create solutions for task
+	}
+
+	public ControllerResult<GenericSolutionModel> check(CheckGenericSolutionDTO checkGenericSolutionDTO) {
+		return null;	//TODO: Check if answer is correct
```
Comment (steve-hb) [inline .docs/solutionTemplates/de/themorpheus/edu/taskservice/controller/solution/GenericSolutionController.java:22]: TODO

File: .docs/solutionTemplates/de/themorpheus/edu/taskservice/database/repository/solution/GenericSolutionRepository.java [hunk 16]
```diff
@@ -0,0 +1,12 @@
+package de.themorpheus.edu.taskservice.database.repository.solution;
+
+import de.themorpheus.edu.taskservice.database.model.solution.GenericSolutionModel;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface GenericSolutionRepository extends JpaRepository<GenericSolutionModel, Integer> {
+
+
```
Comment (steve-hb) [inline .docs/solutionTemplates/de/themorpheus/edu/taskservice/database/repository/solution/GenericSolutionRepository.java]: Only one empty line

File: .docs/solutionTemplates/de/themorpheus/edu/taskservice/endpoint/solution/GenericSolutionEndpoint.java [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class GenericSolutionEndpoint {
+
+	@Autowired
+	GenericSolutionController genericSolutionController;
```
Comment (steve-hb) [inline .docs/solutionTemplates/de/themorpheus/edu/taskservice/endpoint/solution/GenericSolutionEndpoint.java]: Same fkcing line and private

File: .docs/solutionTemplates/de/themorpheus/edu/taskservice/endpoint/solution/GenericSolutionEndpoint.java [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+
+	@PostMapping("/solution/Generic")
```
Comment (steve-hb) [inline .docs/solutionTemplates/de/themorpheus/edu/taskservice/endpoint/solution/GenericSolutionEndpoint.java]: lowercase

File: .gitignore [hunk 19]
```diff
@@ -202,4 +202,5 @@ mvnw.cmd
 # End of https://www.gitignore.io/api/java,maven,eclipse,intellij+all
 /task-microservice-mariadb-data/
 
-.sentryEnabled
\ No newline at end of file
+.sentryEnabled
+test.java
```
Comment (steve-hb) [inline .gitignore]: dafuq?

File: src/main/java/de/themorpheus/edu/taskservice/controller/DifficultyController.java [hunk 20]
```diff
@@ -20,12 +20,12 @@
...
-	public void deleteDifficulty(String nameKey) {
+	public ControllerResult<DifficultyModel> deleteDifficulty(String nameKey) {
 		this.difficultyRepository.deleteDifficultyByNameKeyIgnoreCase(nameKey);
+		return ControllerResult.empty();
 	}
 
 	public ControllerResult<List<DifficultyModel>> getAllDifficulties() {
 		return ControllerResult.of(this.difficultyRepository.findAll());
 	}
-
```
Comment (steve-hb) [inline src/main/java/de/themorpheus/edu/taskservice/controller/DifficultyController.java]: revert

File: src/main/java/de/themorpheus/edu/taskservice/controller/LectureController.java [hunk 21]
```diff
@@ -28,17 +28,16 @@
...
 		this.lectureRepository.deleteLectureByNameKeyIgnoreCase(nameKey);
+		return ControllerResult.empty();
 	}
 
 	public ControllerResult<List<LectureModel>> getAllLecturesFromModule(String moduleNameKey) {
 		ControllerResult<ModuleModel> moduleModel = this.moduleController.getModuleByNameKey(moduleNameKey);
 		if (moduleModel.isResultNotPresent()) return ControllerResult.of(Error.NOT_FOUND, "module");
 		return ControllerResult.of(this.lectureRepository.getLecturesByModuleId(moduleModel.getResult()));
 	}
-
```
Comment (steve-hb) [inline src/main/java/de/themorpheus/edu/taskservice/controller/LectureController.java]: revert

File: src/main/java/de/themorpheus/edu/taskservice/controller/solution/WordsaladSolutionController.java [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+@Component
+public class WordsaladSolutionController {
+
+	private static final String NAME_KEY = "wordsalad_solution";
+
+	@Autowired private SolutionWordsaladRepository solutionWordsaladRepository;
+
+	@Autowired private SolutionController solutionController;
+
+	public ControllerResult<WordsaladSolutionModel> create(CreateWordsaladSolutionDTO dto) {
```
Comment (steve-hb) [inline src/main/java/de/themorpheus/edu/taskservice/controller/solution/WordsaladSolutionController.java]: Im getting headaches

File: src/main/java/de/themorpheus/edu/taskservice/controller/solution/WordsaladSolutionController.java [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+	public ControllerResult<GetWordsaladSolutionDTO> get(int taskId) {
+		ControllerResult<SolutionModel> optionalSolution = this.solutionController.getSolution(taskId, NAME_KEY);
+		if (optionalSolution.isResultNotPresent()) return ControllerResult.ret(optionalSolution);
+
+		Optional<WordsaladSolutionModel> wordsaladSolutionModel = this.solutionWordsaladRepository.findById(
+				optionalSolution.getResult().getSolutionId());
+		if (!wordsaladSolutionModel.isPresent()) return ControllerResult.of(Error.NOT_FOUND, NAME_KEY);
+
+		List<Character> characters = new ArrayList<>();
+		for (char c : wordsaladSolutionModel.get().getSolution().toCharArray()) {
```
Comment (steve-hb) [inline src/main/java/de/themorpheus/edu/taskservice/controller/solution/WordsaladSolutionController.java]: remove brackets

File: src/main/java/de/themorpheus/edu/taskservice/controller/solution/WordsaladSolutionController.java [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+			characters.add(c);
+		}
```
Comment (steve-hb) [inline src/main/java/de/themorpheus/edu/taskservice/controller/solution/WordsaladSolutionController.java]: empty line

File: src/main/java/de/themorpheus/edu/taskservice/endpoint/DifficultyEndpoint.java [hunk 25]
```diff
@@ -44,8 +38,6 @@ public Object getDifficulties() {
 
 	@DeleteMapping("/difficulty/{nameKey}")
 	public Object deleteDifficulty(@PathVariable @NotNull @NotEmpty @NotBlank String nameKey) {
-		this.difficultyController.deleteDifficulty(nameKey);
-		return ControllerResult.empty();
+		return this.difficultyController.deleteDifficulty(nameKey).getHttpResponse();
 	}
-
```
Comment (steve-hb) [inline src/main/java/de/themorpheus/edu/taskservice/endpoint/DifficultyEndpoint.java:50]: revert

File: src/main/java/de/themorpheus/edu/taskservice/endpoint/LectureEndpoint.java [hunk 26]
```diff
@@ -56,5 +47,4 @@ public Object deleteLecture(@PathVariable @NotNull @NotEmpty @NotBlank String na
 		this.lectureController.deleteLecture(nameKey);
 		return ControllerResult.empty();
 	}
-
```
Comment (steve-hb) [inline src/main/java/de/themorpheus/edu/taskservice/endpoint/LectureEndpoint.java]: revert

File: src/main/java/de/themorpheus/edu/taskservice/endpoint/solution/MultipleChoiceSolutionEndpoint.java [hunk 27]
```diff
@@ -0,0 +1,44 @@
...
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class MultipleChoiceSolutionEndpoint {
+
+	@Autowired private MultipleChoiceSolutionController multipleChoiceSolutionController;
+
+	@PostMapping(value = "/solution/multiple_choice", produces = MediaType.APPLICATION_JSON_VALUE, consumes = MediaType.APPLICATION_JSON_VALUE)
+	public Object create(@RequestBody @Valid CreateMultipleChoiceSolutionDTO createMultipleChoiceSolutionDTO) {
```
Comment (steve-hb) [inline src/main/java/de/themorpheus/edu/taskservice/endpoint/solution/MultipleChoiceSolutionEndpoint.java]: ...

File: src/main/java/de/themorpheus/edu/taskservice/endpoint/solution/WordsaladSolutionEndpoint.java [hunk 28]
```diff
@@ -0,0 +1,47 @@
...
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class WordsaladSolutionEndpoint {
+
+	@Autowired private WordsaladSolutionController wordsaladSolutionController;
+
+	@PostMapping("/solution/wordsalad")
+	public Object create(@RequestBody @Valid CreateWordsaladSolutionDTO createWordsaladSolutionDTO) {
```
Comment (steve-hb) [inline src/main/java/de/themorpheus/edu/taskservice/endpoint/solution/WordsaladSolutionEndpoint.java]: ...

File: src/main/java/de/themorpheus/edu/taskservice/endpoint/solution/WordsaladSolutionEndpoint.java [hunk 29]
```diff
@@ -0,0 +1,47 @@
...
+
+	@PutMapping("/solution/wordsalad")
+	public Object update(@RequestBody @Valid UpdateWordsaladSolutionDTO updateWordsaladSolutionDTO) {
+		return this.wordsaladSolutionController.update(updateWordsaladSolutionDTO).getHttpResponse();
+	}
+
+	@DeleteMapping("/solution/wordsalad/{taskId}")
+	public Object delete(@PathVariable @Min(0) int taskId) {
+		return this.wordsaladSolutionController.delete(taskId).getHttpResponse();
+	}
```
Comment (steve-hb) [inline src/main/java/de/themorpheus/edu/taskservice/endpoint/solution/WordsaladSolutionEndpoint.java:46]: missing empty line

**Review by steve-hb — CHANGES_REQUESTED** [review]

File: .docs/solutionTemplates/de/themorpheus/edu/taskservice/controller/solution/GenericSolutionController.java (hunk 15 above)
Comment (PAException) [reply .docs/solutionTemplates/de/themorpheus/edu/taskservice/controller/solution/GenericSolutionController.java:22]: Just the template

File: src/main/java/de/themorpheus/edu/taskservice/controller/solution/WordsaladSolutionController.java (hunk 23 above)
Comment (PAException) [reply src/main/java/de/themorpheus/edu/taskservice/controller/solution/WordsaladSolutionController.java]: which one?

File: src/main/java/de/themorpheus/edu/taskservice/controller/solution/MultipleChoiceSolutionController.java [hunk 30]
```diff
@@ -0,0 +1,82 @@
...
+
+	public ControllerResult<CheckedMultipleChoiceSolutionsDTO> checkMultipleChoiceSolution(CheckMultipleChoiceSolutionDTO dto) {
+		ControllerResult<SolutionModel> optionalSolution = this.solutionController.getSolution(dto.getTaskId(), NAME_KEY);
+		if (optionalSolution.isResultNotPresent()) return ControllerResult.ret(optionalSolution);
+
+		List<MultipleChoiceSolutionModel> multipleChoiceSolutionModels = this.multipleChoiceSolutionRepository.findAllById(
+				Collections.singleton(optionalSolution.getResult().getSolutionId()));
+		if (multipleChoiceSolutionModels.isEmpty()) return ControllerResult.of(Error.NOT_FOUND, NAME_KEY);
+
+		boolean[] checkedSolutions = new boolean[dto.getSolutions().length];
```
Comment (steve-hb) [inline src/main/java/de/themorpheus/edu/taskservice/controller/solution/MultipleChoiceSolutionController.java:44]: Is this in the right order everytime?

**Review by PAException — COMMENTED** [review]
seems good

**Pull request merged**